File: app/src/main/java/com/triangleleft/flashcards/ui/main/DrawerLanguagesAdapter.java
package com.triangleleft.flashcards.ui.main;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.triangleleft.flashcards.R;
import com.triangleleft.flashcards.service.settings.Language;
import com.triangleleft.flashcards.ui.common.FlagImagesProvider;
import com.triangleleft.flashcards.ui.common.OnItemClickListener;

import java.util.Collections;
import java.util.List;

public class DrawerLanguagesAdapter extends RecyclerView.Adapter<DrawerLanguageViewHolder> {

    private final FlagImagesProvider flagImagesProvider;
    private List<Language> languages = Collections.emptyList();
    private OnItemClickListener<DrawerLanguageViewHolder> listener;

    public DrawerLanguagesAdapter(FlagImagesProvider flagImagesProvider,
                                  OnItemClickListener<DrawerLanguageViewHolder> listener) {
        this.flagImagesProvider = flagImagesProvider;
        this.listener = listener;
    }

    public void setData(List<Language> languages) {
        this.languages = languages;
        notifyDataSetChanged();
    }

    @Override
    public DrawerLanguageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_navigation_item, parent, false);
        return new DrawerLanguageViewHolder(view, listener);
    }

    @Override
    public void onBindViewHolder(DrawerLanguageViewHolder holder, int position) {
        Language language = getItem(position);
        holder.showLanguage(flagImagesProvider, language);
    }

    @Override
    public int getItemCount() {
        return languages.size();
    }

    public Language getItem(int position) {
        return languages.get(position);
    }
}

File: app/src/main/java/com/triangleleft/flashcards/ui/main/drawer/DrawerLanguagesAdapter.kt
package com.triangleleft.flashcards.ui.main.drawer

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.triangleleft.flashcards.R
import com.triangleleft.flashcards.service.settings.Language
import com.triangleleft.flashcards.ui.common.FlagImagesProvider
import io.reactivex.subjects.Subject

class DrawerLanguagesAdapter(private val flagImagesProvider: FlagImagesProvider,
                             private val languageClicks: Subject<Language>) : RecyclerView.Adapter<DrawerLanguageViewHolder>() {
    private var languages = emptyList<Language>()

    fun setData(languages: List<Language>) {
        this.languages = languages
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DrawerLanguageViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.view_navigation_item, parent, false)
        return DrawerLanguageViewHolder(view)
    }

    override fun onBindViewHolder(holder: DrawerLanguageViewHolder, position: Int) {
        val language = getItem(position)
        holder.showLanguage(flagImagesProvider, language)
        holder.clicks().subscribe {
            languageClicks.onNext(language)
        }
    }

    override fun getItemCount(): Int {
        return languages.size
    }

    fun getItem(position: Int): Language {
        return languages[position]
    }
}
